File: python/algos/same_trees.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def same_trees(first: Optional[TreeNode], second: Optional[TreeNode]) -> bool:
    if first is None and second is None:
        return True
    elif (first is None and second is not None) or (first is not None and second is None):
        return False
    else:
        return first.val == second.val and same_trees(first.left, second.left) and same_trees(first.right, second.right)


if __name__ == '__main__':
    assert same_trees(TreeNode(2), TreeNode(2))
    assert same_trees(TreeNode(2, TreeNode(1), TreeNode(3)), TreeNode(2, TreeNode(1), TreeNode(3)))
    assert not same_trees(TreeNode(1, TreeNode(1), TreeNode(3)), TreeNode(2, TreeNode(1), TreeNode(3)))

File: rust/src/algos/same_trees.rs
use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

pub fn is_same_tree(
    first: Option<Rc<RefCell<TreeNode>>>,
    second: Option<Rc<RefCell<TreeNode>>>,
) -> bool {
    match (first, second) {
        (None, None) => true,
        (Some(_), None) => false,
        (None, Some(_)) => false,
        (Some(first), Some(second)) => {
            first.borrow().val == second.borrow().val
                && is_same_tree(first.borrow().left.clone(), second.borrow().left.clone())
                && is_same_tree(first.borrow().right.clone(), second.borrow().right.clone())
        }
    }
}
